pkg/sql/pg_catalog.go

```diff
@@ -1835,7 +1843,21 @@ func indexDefFromDescriptor(
 		}
 		indexDef.Interleave = intlDef
 	}
-	fmtCtx := tree.NewFmtCtx(tree.FmtPGIndexDef)
+	if index.IsPartial() {
+		// Format the raw predicate for display in order to resolve user-defined
+		// types to a human readable form.
+		//
+		// TODO(mgartner): Avoid the parsing he predicate expression twice. It
+		// is parsed in schemaexpr.FormatExprForDisplay and again here.
+		formattedPred, err := schemaexpr.FormatExprForDisplayWithoutTypeAnnotations(ctx, table, index.Predicate, p.SemaCtx())
```

Comment:
just a bit confused -- why can't `formattedPred` be used directly as the return value?

Reply:
This is confusing because this function converts an `IndexDescriptor` to a `tree.CreateIndex`. The `tree.CreateIndex.Predicate` field is a `tree.Expr` not a string. So we have to parse it again.

This is definitely not an ideal solution, but I think it's the most minimal one for now. In the long-term, I think it would be prudent to consider whether this conversion back into a `tree.CreateIndex` AST node is the right way to go about this. At the very least, the many code paths for formatting indexes should converge as much as possible.